pkg/asset/machines/gcp/machines.go

```diff
@@ -161,20 +161,27 @@ func provider(clusterID string, platform *gcp.Platform, mpool *gcp.MachinePool,
 	// service accounts with host project privileges. Instead, we can use the existing service account provided
 	// to the installer.
 	if len(platform.NetworkProjectID) > 0 {
-		sess, err := gcpconfig.GetSession(context.TODO())
-		if err != nil {
-			return nil, err
-		}
+		// The service account is only valid for Control Plane Nodes. However, we don't know if this
+		// block is executed for control plane nodes or compute nodes. But validation should have already
+		// occurred to ensure that the service account is not set for compute nodes.
+		if mpool.ServiceAccount != "" {
+			instanceServiceAccount = mpool.ServiceAccount
+		} else {
```

Comment:
I think we need to spell out the logic here a bit more. As this code stands right now, it is still problematic because in the case where `sess.Credentials.JSON` is `nil`, this will still throw an error for the compute nodes. 

My suggestion for an approach would be:
- in all cases, the installler should create the service account for the compute nodes (it seems like this would be fine and not cause any issues with existing installs
- add validation at some point that says if `sess.Credentials.JSON` is `nil` then controlplane.gcp.serviceAccount is required

I think that should work?

Reply:
Is it possible for the `sess.Credentials.JSON` to be `nil`on normal installs or at least those that are not with a shared vpc

Reply:
Yes. Any time the installer is authenticated without the service account key file (so from a VM or the gcloud cli) it will be nil.